src/lib/util.js

```diff
@@ -497,14 +529,27 @@ async function processSpecification(spec, processFunction, args, options) {
 
         // Load the page
         // (note HTTP status is 0 when `file://` URLs are loaded)
+        let cacheInfo;
         if (spec.html) {
             await page.setContent(spec.html, loadOptions);
         }
         else {
+          try {
             const result = await page.goto(spec.url, loadOptions);
             if ((result.status() !== 200) && (!spec.url.startsWith('file://') || (result.status() !== 0))) {
-                throw new Error(`Loading ${spec.url} triggered HTTP status ${result.status()}`);
+              throw new Error(`Loading ${spec.url} triggered HTTP status ${result.status()}`);
+            }
+            const responseHeaders = result.headers();
+            if (responseHeaders.etag) {
+              cacheInfo = {etag: responseHeaders.etag};
+            } else if (responseHeaders['last-modified']) {
+              cacheInfo = {lastModified: responseHeaders['last-modified']};
             }
+          } catch (err) {
+            if (reuseExistingData) {
+              throw new ReuseExistingData(`${spec.url} hasn't changed`);
```

Comment:
I'm not super fan of using exceptions for handling a case that is not going to be exceptional at all. I would prefer if this was handled as part of the regular code flow.

I haven't looked into how easy it may be to take another approach though. There is probably no other way to do it at the network interception level, but once we catch the case here, could we not throw and rather somehow report that the result of processing the spec is that the spec does not need to be processed?

Reply:
I agree using exceptions here is bad, fixed